python/ray/serve/tests/test_grpc.py

```diff
@@ -218,6 +314,418 @@ def __init__(self):
+
+    # Ensures the gRPC Proxy returning the original request id.
+    custom_request_id = "fake-request-id"
+    metadata = (("request_id", custom_request_id),)
+    response, call = stub.__call__.with_call(request=request, metadata=metadata)
+    response_request_id = None
+    for key, value in call.trailing_metadata():
+        if key == "request_id":
+            response_request_id = value
+            break
+    assert custom_request_id == response_request_id
+
+    # Ensures the gRPC Proxy returning a new request id.
+    response, call = stub.__call__.with_call(request=request)
+    response_request_id = None
+    for key, value in call.trailing_metadata():
+        if key == "request_id":
+            response_request_id = value
+            break
+    assert custom_request_id != response_request_id
+
+
+def test_grpc_proxy_on_draining_nodes(ray_cluster):
+    """Test gRPC request on the draining node.
+
+    When there are no replicas on head node and some replicas on the worker node, the
+    ListApplications and Healthz methods should respond successfully. When there are
+    no replicas on any nodes, ListApplications and Healthz methods should continue to
+    succeeding on the head node. But should return draining response on the worker node.
+    """
+    head_node_grpc_port = 9000
+    worker_node_grpc_port = 9001
+
+    # Setup worker gRPC proxy to be pointing to port 9001. Head node gRPC proxy will
+    # continue to be pointing to the default port 9000.
+    os.environ["TEST_WORKER_NODE_GRPC_PORT"] = str(worker_node_grpc_port)
+
+    # Set up a cluster with 2 nodes.
+    cluster = Cluster()
+    cluster.add_node(num_cpus=0)
+    cluster.add_node(num_cpus=2)
+    cluster.wait_for_nodes()
+    ray.init(address=cluster.address)
+
+    # Start serve with gRPC proxy
+    grpc_servicer_functions = [
+        "ray.serve.generated.serve_pb2_grpc.add_UserDefinedServiceServicer_to_server",
+        "ray.serve.generated.serve_pb2_grpc.add_FruitServiceServicer_to_server",
+    ]
+    serve.start(
+        http_options={"location": "EveryNode"},
+        grpc_options={
+            "port": head_node_grpc_port,
+            "grpc_servicer_functions": grpc_servicer_functions,
+        },
+    )
+
+    # Deploy 2 replicas, both should be on the worker node.
+    @serve.deployment(num_replicas=2)
+    class HelloModel:
+        def __call__(self):
+            return serve_pb2.UserDefinedResponse(greeting="hello")
+
+    model = HelloModel.bind()
+    app_name = "app1"
+    serve.run(target=model, name=app_name)
+
+    # Ensure worker node has both replicas.
+    def check_replicas_on_worker_nodes():
+        _actors = ray._private.state.actors().values()
+        replica_nodes = [
+            a["Address"]["NodeID"]
+            for a in _actors
+            if a["ActorClassName"].startswith("ServeReplica")
+        ]
+        return len(set(replica_nodes)) == 1
+
+    wait_for_condition(check_replicas_on_worker_nodes)
```

Comment:
Would be a good idea to set a timeout to make sure test wouldn't hang

Reply:
there is already a default timeout of 10 seconds :)